Adding attribute distinct cardinalities using HyperLogLog statistics, #1296

jarohen opened the pull request:
resolves #1296 

We now keep [HyperLogLog](https://en.wikipedia.org/wiki/HyperLogLog) buffers for each attribute on ingestion so that we know, for each attribute, how many distinct values and distinct entity ids we have. This is predominantly to fix the query join order in the case where two attributes have similar doc-counts but very different selectivities. Previously, when a user is filtering both by `:type` (say only a handful of different types) and `:name` (assuming mostly unique between entities), Crux wouldn't necessarily choose to filter first by `:name`. See `picks-more-selective-join-order` test. 

See https://github.com/orgs/juxt/teams/crux-core/discussions/22

Benchmarks-wise, this currently adds 4-5% to ingest (makes sense, we're doing more work, although we may be able to shave off some time), and no appreciable difference in Watdiv nor TPC-H queries.
* NOTE - we could consider caching the stats KVs between transactions so that we don't have to go look them up. This may not be that much of a performance improvement given Rocks et al will likely cache recently changed/accessed KVs in memory. In practice, it seems that the hashing of the values for the HLL dominates this time.
* NOTE - the cache key for `compile-sub-query` doesn't take the relative attribute cardinalities into account. if these change significantly, the cache won't get invalidated, and we may have sub-optimal queries.

hraberg committed:
Import Håkan's HLL sketch

jarohen committed:
reduce fn names, don't need to specify `hyper-log-log-` in each because they'll be called as `hll/estimate` etc

jarohen committed:
checkpoint: statistics are the domain of the index-store

jarohen committed:
storing approximate cardinalities in the attribute stats :)

jarohen committed:
example TPC-H code -> fixtures.tpch

jarohen committed:
keep both value and eid cardinalities, and update query planner

jarohen committed:
test that shows HLL affects the join order

hraberg commented on crux-core/src/crux/fork.clj:
```diff
@@ -150,6 +155,22 @@
+               (db/all-attrs transient-index-snapshot)))
+
+  (doc-count [_ attr]
+    (or (db/doc-count transient-index-snapshot attr)
```
Think this does want to be addition?

jarohen replied:
Don't _think_ so - the attribute HLL gets copied into the transient store, so it completely replaces the one in the persistent store?

hraberg replied:
Ah, ignore me then, didn't follow it through, just assumed that the transient one was an empty, local one.

hraberg commented on crux-core/src/crux/fork.clj:
```diff
@@ -150,6 +155,22 @@
+        (db/doc-count persistent-index-snapshot attr)))
+
+  (value-cardinality [_ attr]
+    (or (db/value-cardinality transient-index-snapshot attr)
```
This should in theory be `hll/estimate-union` but as that is likely hidden at this level, it can maybe be `max`?

jarohen replied:
likewise?

hraberg commented on crux-core/src/crux/query.clj:
```diff
@@ -1388,7 +1387,7 @@
                                                                        type->clauses
                                                                        in-vars
-                                                                       stats)
+                                                                       db)
```
Bit unsure here, so likely fine, but the entire idea of the encode-value-fn is to clearly separate compile time from runtime of the query, and not use the index-snapshot during the former.

jarohen replied:
Ah, ok, thanks - seems a quick chat will be best to resolve this one?

jarohen replied:
TODO: hide this behind a wrapper for now so that it's obvious that the HLL provider shouldn't be passed through to query runtime.

jarohen replied:
done :smile:

hraberg reviewed (commented):
LGTM on a glance, I would probably test caching of the HLL to see what, if any, benefit it has during ingest.

Also, see note about the compile time  / runtime separation, not clear to me if there's any real issue here the way its used.

An idea I had earlier was to cache the relative cardinalities used in the query as part of the cache key, so if they change, the query plan would get regenerated. But that's not how it works now, so not part of this PR.

jarohen committed:
hiding stats fns behind a different protocol so that we don't risk the DB crossing the planning/execution boundary

it still closes over the index-snapshot, but hopefully passing the stats obj over the boundary will be more obvious
